docker-compose-rule-junit4/src/main/java/com/palantir/docker/compose/EventEmitter.java

```diff
@@ -0,0 +1,162 @@
+    }
+
+    private Consumer<Cluster> clusterWait(
+            ClusterWaitType clusterWaitType,
+            ClusterWait clusterWait) {
+        // This weird bit of complexity is because we can't tell what services a ClusterWait is using until it
+        // actually runs. So we have to record the services it accesses then use this to generate the events. The
+        // Optional exists solely as a check again logic errors - in the case when events are generated before the
+        // cluster wait has begun.
+        AtomicReference<Optional<Set<String>>> recordedServiceNames = new AtomicReference<>(Optional.empty());
+
+        Consumer<Cluster> recordingClusterWait = cluster -> {
+            RecordingCluster recordingCluster = new RecordingCluster(cluster);
+            try {
+                clusterWait.waitUntilReady(recordingCluster);
+            } finally {
+                recordedServiceNames.set(Optional.of(recordingCluster.recordedContainerNames()));
+            }
+        };
+
+        return cluster -> emitNotThrowing(
+                () -> recordingClusterWait.accept(cluster),
+                task -> Event.clusterWait(ClusterWaitEvent.builder()
+                        .task(task)
+                        .serviceNames(recordedServiceNames.get().orElseThrow(
+                                () -> new IllegalStateException("Recorded service names have not yet been computed")))
+                        .type(clusterWaitType)
+                        .build()));
+    }
+
+    private void emitNotThrowing(CheckedRunnable runnable, Function<Task, Event> eventFunction) {
+        try {
+            emitTask(runnable, eventFunction);
+        } catch (InterruptedException | IOException e) {
+            Throwables.propagate(e);
+        }
+    }
+
+    private void emitTask(CheckedRunnable runnable, Function<Task, Event> eventFunction)
+            throws IOException, InterruptedException {
+        Optional<String> failure = Optional.empty();
+        OffsetDateTime startTime = clock.instant().atOffset(ZoneOffset.UTC);
+        try {
+            runnable.run();
+        } catch (RuntimeException | IOException | InterruptedException e) {
+            failure = Optional.of(Exceptions.condensedStacktraceFor(e));
+            throw e;
+        } finally {
+            OffsetDateTime endTime = clock.instant().atOffset(ZoneOffset.UTC);
+            Task task = Task.builder()
+                    .startTime(startTime)
+                    .endTime(endTime)
+                    .failure(failure)
+                    .build();
+
+            Event event = eventFunction.apply(task);
+            emitEvent(event);
+        }
+    }
+
+    private void emitEvent(Event event) {
+        eventConsumers.forEach(eventConsumer -> {
+            try {
+                eventConsumer.receiveEvent(event);
+            } catch (Exception e) {
+                log.error("Error sending event {}", event, e);
```

Comment:
I think we should propagate the error if an eventConsumer fails

Reply:
What happens if you have multiple event consumers? Then if one fails to consume an event the rest don't get the event.

Reply:
I'm trying to avoid unexpected behaviour when a user provides their own consumers. Its kinda weird, but I could imagine users adding event consumers to ensure some sort of preconditions, and failing the test (suite|case) if they do not hold

Reply:
Ok, this will now throw and add all exceptions that happened as suppressed.